### PR TITLE
Support iostat in read()

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
This PR gets the chat program in the lf33run branch at fastgpt (https://github.com/certik/fastGPT/tree/lf33run) to run. It currently works for first 2 prompts but segfaults on the exact third run. Also, note that in the lf33run branch, to make it work with LFortran, we replaced a local array allocated on stack with an allocatable array allocated on heap.

### CODE REVIEW

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -6661,6 +6662,12 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
             this->visit_expr_wrapper(x.m_unit, true);
             unit_val = tmp;
         }
+        if (x.m_iostat) {
+            this->visit_expr_wrapper(x.m_iostat, true);
+            iostat = builder->CreateIntToPtr(tmp, llvm_utils->getIntType(4, true));
+        } else {
+            iostat = llvm::Constant::getNullValue(llvm_utils->getIntType(4, true));
+        }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_llvm.cpp:6670]: This leads to a segmentation fault on call to `_lfortran_read_char()` from `llvm`. Is this the correct way to pass a pointer from llvm to C?

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 2]
```diff
@@ -1937,12 +1937,16 @@ LFORTRAN_API void _lfortran_read_array_int32(int32_t *p, int array_size, int32_t
...
+LFORTRAN_API void _lfortran_read_char(char **p, int32_t unit_num, int32_t *iostat)
 {
     if (unit_num == -1) {
         // Read from stdin
-        *p = (char*)malloc(strlen(*p) * sizeof(char));
-        scanf("%s", *p);
+        int n = strlen(*p);
+        *p = (char*)malloc(n * sizeof(char));
+        if (fgets(*p, n, stdin)) {
+            *iostat = 0;
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:1947]: At this line `*iostat = 0`, are we allowed to set the value of `iostat` to `0` here. It seems iostat is in integer variable in the user source code which has its memory allocated in llvm stack. Is C allowed to access/update/write to a memory used by llvm?

It seems in the `lfortran_intrinsics.c` at any other place where we have such type of memory update/write from C to LLVM, the memory used is allocated in C heap. In the above case, specifically we have memory allocated in LLVM stack and is being written/updated by C.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (certik) [reply src/libasr/codegen/asr_to_llvm.cpp:6670]: I am not sure --- I think we are passing pointers elsewhere, so just see if this is consistent.

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 2 above)
Comment (certik) [reply src/libasr/runtime/lfortran_intrinsics.c:1947]: Yes, you can update LLVM allocated memory from C. Obviously the pointer to it must not be dangling, but otherwise it should work.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 3]
```diff
@@ -6661,6 +6662,12 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
             this->visit_expr_wrapper(x.m_unit, true);
             unit_val = tmp;
         }
+        if (x.m_iostat) {
+            this->visit_expr_wrapper(x.m_iostat, true);
+            iostat = builder->CreateIntToPtr(tmp, llvm_utils->getIntType(4, true));
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:6667]:
```suggestion
        if (x.m_iostat) {
            uint64_t ptr_loads_copy = ptr_loads;
            ptr_loads = 0;
            this->visit_expr_wrapper(x.m_iostat, false);
            ptr_loads = ptr_loads_copy;
            iostat = tmp;
```

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

**Comment by czgdp1807** [comment]
What you are doing is giving `iostat` value a pointer representation. What's needed is to write the new value to `iostat` (i.e., pass the address of `iostat` to C and not the value of `iostat` as an address).

**Comment by certik** [comment]
This PR is fine, but I have two questions:

* What bug does it fix? Let's add a test for it? Maybe reading some stuff from an empty file?
* The change in fastGPT deviates from master. Does this workaround make chat work?

**Comment by certik** [comment]
@Shaikh-Ubaid do you know what changes to make so that test_chat works as well? Probably a similar allocatable change?

**Comment by Thirumalai-Shaktivel** [comment]
test_chat seems to work for me:
```console
$ ./test_chat
Your name is fastGPT and you are an AI bot. The user will ask you questions and you answer in a nice, truthful, short way.
User: What is the capital of Czechia?
fastGPT: Prague.
User: How many legs does a dog have?
fastGPT: Four.
User: What color does the sky have?
fastGPT: Blue.
User: What can you type a document on?
fastGPT: A document.
User: What can you drive in?
fastGPT: A car.
User:
```

**Comment by Thirumalai-Shaktivel** [comment]
Chat works as well!
I think we have it, everything works!

**Comment by certik** [comment]
The test_chat must be tested at the CI, and then just "ctest" should run. For me it doesn't run.

**Comment by certik** [comment]
@Thirumalai-Shaktivel can you submit a PR to test this at the CI?

**Comment by certik** [comment]
Ok, perfect! Now it makes sense.

**Comment by Shaikh-Ubaid** [comment]
>What bug does it fix? Let's add a test for it? Maybe reading some stuff from an empty file?

Sure, I will add test for it.

>The change in fastGPT deviates from master. Does this workaround make chat work?

Yes, it does partially make the chat work.

>do you know what changes to make so that test_chat works as well? Probably a similar allocatable change?

I think it fails due to https://github.com/lfortran/lfortran/issues/2360. I will fix it.
